flamingo/common.h

```diff
@@ -50,50 +50,303 @@ static inline int find_static_members_in_class(flamingo_t* flamingo, flamingo_sc
+ *
+ * @return The new scope.
+ */
 static inline flamingo_scope_t* scope_alloc(void);
+
+/**
+ * Free a scope.
+ *
+ * This frees the scope and all its variables. It should only be called when the reference count reaches 0.
+ *
+ * @param scope The scope to free.
+ */
 static inline void scope_free(flamingo_scope_t* scope);
+
+/**
+ * Decrement the reference count of a scope.
+ *
+ * If the reference count reaches 0, the scope is freed.
+ *
+ * @param scope The scope to decrement the reference count of. Can be NULL.
+ */
 static inline void scope_decref(flamingo_scope_t* scope);
+
+/**
+ * Add a variable to a scope.
+ *
+ * @param scope The scope to add the variable to.
+ * @param key The name of the variable.
+ * @param key_size The size of the variable name.
+ * @return The new variable.
+ */
 static inline flamingo_var_t* scope_add_var(flamingo_scope_t* scope, char const* key, size_t key_size);
+
+/**
+ * Find a variable in a scope (shallow search).
+ *
+ * This only searches the given scope, not any parent scopes.
+ *
+ * @param scope The scope to search.
+ * @param key The name of the variable to find.
+ * @param key_size The size of the variable name.
+ * @return The variable if found, or NULL if not found.
+ */
 static inline flamingo_var_t* scope_shallow_find_var(flamingo_scope_t* scope, char const* key, size_t key_size);
 
 // Variable prototypes.
 
+/**
+ * Set the value of a variable.
+ *
+ * This also updates the value's name to match the variable's key.
+ *
+ * @param var The variable to set the value of.
+ * @param val The value to set. Can be NULL.
+ */
 static inline void var_set_val(flamingo_var_t* var, flamingo_val_t* val);
 
 // Value prototypes.
 
+/**
+ * Increment the reference count of a value.
+ *
+ * @param val The value to increment the reference count of.
+ * @return The value itself.
+ */
 static inline flamingo_val_t* val_incref(flamingo_val_t* val);
+
+/**
+ * Initialize a value.
+ *
+ * Sets the value to be an anonymous NONE value with a reference count of 1.
+ *
+ * @param val The value to initialize.
+ * @return The initialized value.
+ */
 static inline flamingo_val_t* val_init(flamingo_val_t* val);
+
+/**
+ * Get the string representation of a value's type.
+ *
+ * @param val The value to get the type string of.
+ * @return The string representation of the value's type (e.g. "integer", "string", "vector").
+ */
 static inline char const* val_type_str(flamingo_val_t const* val);
+
+/**
+ * Get the string representation of a value's role.
+ *
+ * @param val The value to get the role string of.
+ * @return The string representation of the value's role (e.g. "variable", "function", "class").
+ */
 static inline char const* val_role_str(flamingo_val_t* val);
+
+/**
+ * Allocate a new value.
+ *
+ * The value is initialized with `val_init`.
+ *
+ * @return The new value.
+ */
 static inline flamingo_val_t* val_alloc(void);
+
+/**
+ * Create a deep copy of a value.
```

Comment:
perhaps define "deep" in description? Like vectors inside of vectors will be copied too.